res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,1949 @@
+        this.isLongPressed = false;
+        this.longPressTimer = 0;
+    }
+    input(pressed) {
+        if (pressed) {
+            script.toggleControl(this.group, this.inKey);
+            this.longPressTimer = engine.beginTimer(this.longPressTimeOut, () => {
+                this.isLongPressed = true;
+                this.longPressTimer = 0;
+            }, true);
+        } else {
+            if (this.isLongPressed) {
+                script.toggleControl(this.group, this.inKey);
+            }
+            if (this.longPressTimer !== 0) {
+                engine.stopTimer(this.longPressTimer);
+            }
+            this.longPressTimer = 0;
+            this.isLongPressed = false;
+        }
+    }
+}
+
+class PlayButton extends ToggleButton {
+    constructor(options) {
+        super(options);
+        this.inKey = "play";
+        this.outKey = "play_indicator";
+        this.outConnect();
+    }
+}
+
+class CueButton extends PushButton {
+    constructor(options) {
+        super(options);
+        this.outKey = "cue_indicator";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = "cue_default";
+    }
+    shift() {
+        this.inKey = "start_stop";
+    }
+}
+
+class Encoder extends Component {
+    constructor(options) {
+        super(options);
+        this.lastValue = null;
+    }
+    isRightTurn(value) {
+        // detect wrap around
+        const oldValue = this.lastValue;
+        this.lastValue = value;
+        if (oldValue === this.max && value === 0) {
+            return true;
+        }
+        if (oldValue === 0 && value === this.max) {
+            return false;
+        }
+        return value > oldValue;
+    }
+}
+
+class HotcueButton extends PushButton {
+    constructor(options) {
+        super(options);
+        if (this.number === undefined || !Number.isInteger(this.number) || this.number < 1 || this.number > 32) {
+            throw Error("HotcueButton must have a number property of an integer between 1 and 32");
+        }
+        this.outKey = "hotcue_" + this.number + "_enabled";
+        this.colorKey = "hotcue_" + this.number + "_color";
+        this.outConnect();
+    }
+    unshift() {
+        this.inKey = "hotcue_" + this.number + "_activate";
+    }
+    shift() {
+        this.inKey = "hotcue_" + this.number + "_clear";
+    }
+    output(value) {
+        if (value) {
+            this.send(this.color + this.brightnessOn);
+        } else {
+            this.send(0);
+        }
+    }
+    outConnect() {
+        if (undefined !== this.group) {
+            this.outConnections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+            this.outConnections[1] = engine.makeConnection(this.group, this.colorKey, (colorCode) => {
+                this.color = this.colorMap.getValueForNearestColor(colorCode);
+                this.output(engine.getValue(this.group, this.outKey));
+            });
+        }
+    }
```

Comment:
Can you elaborate?

Reply:
Well thats largely related to my previous comment that this only supports half the usecase that the ColoredHotcueButton was actually made for.

Reply:
but a quick example that comes to mind is the `this.color + this.brightnessOn` part because thats device specific. 
Also `this.send(0);` should rather be `this.send(this.off);`